cmd/vic-machine/ops-user/rbac.go

```diff
@@ -0,0 +1,286 @@
+	mgr := &AuthzManager{
+		session:      session,
+		authzManager: authManager,
+		principal:    principal,
+	}
+	return mgr
+}
+
+func (am *AuthzManager) CreateRoles(ctx context.Context) (int, error) {
+	return am.createOrRepairRoles(ctx, getTargetRoles())
+}
+
+func (am *AuthzManager) DeleteRoles(ctx context.Context) (int, error) {
+	return am.deleteRoles(ctx, getTargetRoles())
+}
+
+func (am *AuthzManager) RoleList(ctx context.Context) (object.AuthorizationRoleList, error) {
+	return am.getRoleList(ctx)
+}
+
+func (am *AuthzManager) createOrRepairRoles(ctx context.Context, targetRoles []types.AuthorizationRole) (int, error) {
+	// Get all the existing roles
+	mgr := am.authzManager
+	roleList, err := mgr.RoleList(ctx)
+
+	if err != nil {
+		return 0, err
+	}
+
+	var count int
+	for _, targetRole := range targetRoles {
+		foundRole := roleList.ByName(targetRole.Name)
+		if foundRole != nil {
+			isMod, err := am.checkAndRepairRole(ctx, &targetRole, foundRole)
+			if isMod && err == nil {
+				count++
+			}
+		} else {
+			_, err = mgr.AddRole(ctx, targetRole.Name, targetRole.Privilege)
+			count++
+		}
+		if err != nil {
+			return count, err
+		}
+	}
+	return count, nil
+}
+
+func (am *AuthzManager) deleteRoles(ctx context.Context, targetRoles []types.AuthorizationRole) (int, error) {
+	mgr := am.authzManager
+	// Get all the existing roles
+	roleList, err := mgr.RoleList(ctx)
+
+	if err != nil {
+		return 0, err
+	}
+
+	var count int
+	for _, targetRole := range targetRoles {
+		foundRole := roleList.ByName(targetRole.Name)
+		if foundRole != nil {
+			err = mgr.RemoveRole(ctx, foundRole.RoleId, true)
+			if err == nil {
+				count++
+			}
+		}
+	}
+	return count, nil
+}
+
+func (am *AuthzManager) getRoleList(ctx context.Context) (object.AuthorizationRoleList, error) {
+	return am.authzManager.RoleList(ctx)
+}
+
+func (am *AuthzManager) checkAndRepairRole(ctx context.Context,
+	tRole *types.AuthorizationRole, fRole *types.AuthorizationRole) (bool, error) {
+	mgr := am.authzManager
+	// Check that the privileges list in Target Role is a subset of the list in Found role
+	fSet := make(map[string]bool)
+	for _, p := range fRole.Privilege {
+		fSet[p] = true
+	}
+
+	var isModified bool
+	for _, p := range tRole.Privilege {
+		if _, found := fSet[p]; !found {
+			// Privilege not found
+			// Add it to the found Role
+			fRole.Privilege = append(fRole.Privilege, p)
+			isModified = true
+		}
+	}
+
+	if !isModified {
+		return false, nil
+	}
+
+	// Not a subset need to call go-vmomi to set the new privileges
+	err := mgr.UpdateRole(ctx, fRole.RoleId, fRole.Name, fRole.Privilege)
+
+	return true, err
+}
+
+func getTargetRoles() []types.AuthorizationRole {
+	count := len(RBACConf.Resources)
+	roles := make([]types.AuthorizationRole, 0)
+	dSet := make(map[string]bool)
+	for i := 0; i < count; i++ {
+		name := rolePrefix + RBACConf.Resources[i].Role.Name
+		// Discard duplicates
+		if _, found := dSet[name]; !found {
```

Comment:
Not sure why you're discarding duplicates given it's an inline struct - should we `panic` instead as it seems like it's programmer error in this circumstance?

Reply:
In the inline structure there are roles that legitimately appear in multiple entries. For instance, the Datastore Role applies to Cluster, Datastore and SANDatastore. This method returns the set of roles (no duplication).